packages/mui-material/src/Modal/Modal.js

```diff
@@ -102,19 +102,23 @@ const Modal = React.forwardRef(function Modal(inProps, ref) {
 
   const classes = extendUtilityClasses(ownerState);
 
-  const Root = components.Root ?? component ?? ModalRoot;
+  const Root = components.Root ?? component;
```

Comment:
I'm a bit out of the loop with the way props flows now, but I don't get why we would need this
```suggestion
```

Would this be enough?

```diff
diff --git a/packages/mui-material/src/Modal/Modal.js b/packages/mui-material/src/Modal/Modal.js
index d631549e94..714bc3e6ef 100644
--- a/packages/mui-material/src/Modal/Modal.js
+++ b/packages/mui-material/src/Modal/Modal.js
@@ -5,6 +5,7 @@ import { isHostComponent, resolveComponentProps } from '@mui/base/utils';
 import { elementAcceptingRef, HTMLElementType } from '@mui/utils';
 import styled from '../styles/styled';
 import useThemeProps from '../styles/useThemeProps';
+import shouldSpreadAdditionalProps from '../utils/shouldSpreadAdditionalProps';
 import Backdrop from '../Backdrop';

 export const modalClasses = modalUnstyledClasses;
@@ -102,8 +103,6 @@ const Modal = React.forwardRef(function Modal(inProps, ref) {

   const classes = extendUtilityClasses(ownerState);

-  const Root = components.Root ?? component;
-
   return (
     <ModalUnstyled
       components={{
@@ -112,18 +111,17 @@ const Modal = React.forwardRef(function Modal(inProps, ref) {
         ...components,
       }}
       componentsProps={{
-        root: () => ({
+        root: {
           ...resolveComponentProps(componentsProps.root, ownerState),
-          ...(!isHostComponent(Root) && { theme }),
-          /** We pass the 'as' prop always because components.Root is set as ModalRoot wrapped with the
-           * styled API which thus supports the 'as' prop for host or React components.
-           */
-          as: Root,
-        }),
-        backdrop: () => ({
+          ...(shouldSpreadAdditionalProps(components.Root) && {
+            as: component,
+            ownerState: { ...componentsProps.root?.ownerState, },
+          }),
+        },
+        backdrop: {
           ...BackdropProps,
           ...resolveComponentProps(componentsProps.backdrop, ownerState),
-        }),
+        },
       }}
       onTransitionEnter={() => setExited(false)}
       onTransitionExited={() => setExited(true)}
```

Reply:
> I'm a bit out of the loop with the way props flows now, but I don't get why we would need this
 
A custom Root component can be passed through the slot or `component` prop.

> Would this be enough?

It works, the current tests do pass with your suggested change. Although I am not sure about this line `ownerState: { ...componentsProps.root?.ownerState, },` in the diff as `componentsProps.root` can be a callback or an object and whether `useSlotProps` inside `ModalUnstyled` already appends ownerState properly.

What's the advantage of this?

IMO, I would rather keep my approach as it is simpler and easy to read at a first glance.

Instead what I have done is remove the unnecessary function definition from `componentsProps.root` and `componentsProps.backdrop` in commit https://github.com/mui/material-ui/pull/33934/commits/adb6d4e6864da4e43016d7c27969e928dafd78f2 after seeing your suggested change.

Reply:
> What's the advantage of this?

I guess it's simpler and it matches how it's done in https://github.com/mui/material-ui/blob/dee9529f7a298c54ae760761112c3ae9ba082137/packages/mui-material/src/Slider/Slider.js#L518

AFAIK, we should never provide `components.Root` for the `as` prop. `components.Root` is meant to replace the first level component, while the `component` prop is meant to replace the leaf component.

Reply:
> I guess it's simpler and it matches how it's done in:

Okay but IMO, I think it is not simple compared to what I did. As for it being consistent with other components what about the following line:
```
ownerState: { ...componentsProps.root?.ownerState, }
```
This is wrong as `componentProps` slots can be either a callback or an object as I said in the previous comment. If we keep it by also considering callback here, we need to change in other components later on.

> components.Root is meant to replace the first level deep component, while the component prop is meant to replace the leaf component

Can you please elaborate on this? I think I understand what you mean but the `as` prop can also be used to pass custom React components. Also `component` prop can be used for React components not only for overriding HTML elements ([docs](https://mui.com/base/getting-started/usage/#components)).

Reply:
@oliviertassinari This matches the implementation of Base components. In this library `component` = `components.Root`. I don't understand what you mean by "components.Root is meant to replace the first level deep component, while the component prop is meant to replace the leaf component" - which leaf component exactly?

Reply:
@ZeeshanTamboli, as discussed in #34334, the `component` prop should be passed into the `as` prop, while the `components.Root` should be passed into ModalUnstyled's `components.Root`. This way developers can either customize the whole slot or the ultimately rendered component